activerecord/lib/active_record/optimistic_update.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module OptimisticUpdate
+    def update_if(*args, &block)
+      @_optimistic_update_constraints << args
+
+      if block_given?
+        begin
+          yield
+        ensure
+          clear_update_conditions
+        end
+      else
+        self
+      end
+    end
+
+    def clear_update_conditions
+      @_optimistic_update_constraints = []
+    end
+
+    def decrement!(...)
+      super(...)
+    rescue ActiveRecord::UnexpectedValueInDatabase => e
+      raise ActiveRecord::UnexpectedValueInDatabase.new(self, "decrement!", e.constraints)
+    end
+
+    def update_columns(attributes)
+      return super if @_optimistic_update_constraints.empty?
+
+      _with_scope do |scope|
+        super.tap do |success|
+          raise ActiveRecord::UnexpectedValueInDatabase.new(self, "update", scope.arel.where_sql) unless success
+        end
+      end
+    end
+
+    private
+      def init_internals
+        super
+        @_optimistic_update_constraints = []
+      end
+
+      def _update_row(attribute_names, attempted_action = "update")
```

Comment:
I'm not sure what underscore prefix convention was for internal private methods, I used them here, but I think I saw some internal private methods in `Persistence` that didn't have it, so wasn't sure what the recommended approach was.

Also, not sure when `:nodoc:` should be used, is it for public methods only and is not necessary for private methods?